File: sdk/typescript/src/providers/gemini.ts
/**
 * Gemini provider wrapping for MemG.
 *
 * Supports three modes:
 * - **native**: Full in-process engine (no Go server needed).
 * - **client**: Intercepts calls locally, querying MCP for memory context.
 * - **proxy**: Not supported for Gemini (Gemini SDK does not use OpenAI-compatible endpoints).
 */

import { MemGClient } from '../client';
import { wrapGeminiClient } from '../intercept';
import type { WrapOptions } from '../types';

/**
 * Wrap a Gemini GenerativeModel with MemG memory capabilities.
 */
export function wrap(client: any, opts: WrapOptions): any {
  if (opts.mode === 'native') {
    return wrapGeminiNative(client, opts);
  }
  if (opts.mode === 'client') {
    const mcp = new MemGClient(opts.mcpUrl);
    return wrapGeminiClient(client, mcp, opts.entity, opts.extract ?? true);
  }
  // Proxy mode: Gemini SDK doesn't route through OpenAI-compatible proxy.
  // Fall through to native mode as the best alternative.
  console.warn('[memg] Proxy mode is not supported for Gemini clients. Using native mode instead.');
  return wrapGeminiNative(client, opts);
}

function wrapGeminiNative(client: any, opts: WrapOptions): any {
  const { MemG } = require('../index') as typeof import('../index');

  const memg = new MemG(opts.nativeConfig);
  let initPromise: Promise<void> | null = null;

  async function ensureInit(): Promise<typeof memg> {
    if (!initPromise) {
      initPromise = memg.init();
    }
    await initPromise;
    return memg;
  }

  const originalGenerate = client.generateContent.bind(client);

  client.generateContent = async function patchedGenerate(
    params: any
  ): Promise<any> {
    const m = await ensureInit();
    const entityId = opts.entity || 'default';

    // Normalize string input to Gemini format.
    let augmentedParams =
      typeof params === 'string'
        ? { contents: [{ role: 'user', parts: [{ text: params }] }] }
        : { ...params };

    // Step 1: Build memory context and inject into systemInstruction.
    try {
      const userText = extractLastGeminiUserMessage(augmentedParams);
      if (userText) {
        const context = await m.buildMemoryContext(entityId, userText);
        if (context) {
          augmentedParams = injectGeminiContext(augmentedParams, context);
        }
      }
    } catch (err) {
      console.warn('[memg] Failed to build memory context, proceeding without:', err);
    }

    // Step 2: Call the original generateContent.
    const response = await originalGenerate(augmentedParams);

    // Step 3: Extract knowledge from the exchange (fire-and-forget).
    if (opts.extract !== false) {
      try {
        const contents = augmentedParams.contents || [];
        const messages = contents
          .filter((c: any) => c.role === 'user' || c.role === 'model')
          .map((c: any) => ({
            role: c.role === 'model' ? 'assistant' : c.role,
            content: c.parts?.map((p: any) => p.text).join('') ?? '',
          }));

        const assistantText =
          response?.response?.candidates?.[0]?.content?.parts?.[0]?.text ??
          response?.response?.text?.();
        if (assistantText) {
          messages.push({ role: 'assistant', content: assistantText });
        }

        m.extractFromMessages(entityId, messages).catch(() => {});
      } catch {
        // Never let extraction errors affect the response.
      }
    }

    return response;
  };

  // Attach close method so users can clean up.
  client._memg = memg;
  client._memgClose = () => memg.close();

  return client;
}

function extractLastGeminiUserMessage(params: any): string | null {
  const contents = params?.contents;
  if (!Array.isArray(contents)) return null;
  for (let i = contents.length - 1; i >= 0; i--) {
    if (contents[i].role === 'user' && contents[i].parts?.length > 0) {
      return contents[i].parts[0].text ?? null;
    }
  }
  return null;
}

function injectGeminiContext(params: any, contextText: string): any {
  const result = { ...params };
  const existing = result.systemInstruction;
  if (existing && existing.parts?.length > 0) {
    const existingText = existing.parts.map((p: any) => p.text).join('');
    result.systemInstruction = {
      parts: [{ text: existingText + '\n\n' + contextText }],
    };
  } else {
    result.systemInstruction = { parts: [{ text: contextText }] };
  }
  return result;
}

File: sdk/python/memg/providers/gemini.py
"""Gemini provider wrapping for MemG.

Supports three modes:
- native: full in-process engine
- client: intercepts calls locally via MCP
- proxy: not supported for Gemini (falls back to native)
"""

from __future__ import annotations

import logging
import threading
from typing import Any, Optional

logger = logging.getLogger("memg")


def wrap(
    client: Any,
    entity: Optional[str] = None,
    mode: str = "native",
    proxy_url: str = "http://localhost:8787/v1",
    mcp_url: str = "http://localhost:8686",
    extract: bool = True,
    config: Any = None,
    embedder: Any = None,
    **kwargs: Any,
) -> Any:
    """Wrap a Gemini GenerativeModel with MemG memory."""
    if mode == "client":
        from ..client import MemGClient
        from ..intercept import wrap_gemini_client
        mcp_client = MemGClient(mcp_url)
        return wrap_gemini_client(client, mcp_client, entity or "default", extract)

    if mode == "proxy":
        logger.warning("memg: proxy mode not supported for Gemini, using native mode")

    return _wrap_native(client, entity, extract, config, embedder, **kwargs)


def _wrap_native(
    client: Any,
    entity: Optional[str],
    extract: bool,
    config: Any,
    embedder: Any,
    **kwargs: Any,
) -> Any:
    """Wrap with the native in-process engine."""
    from .. import MemG as MemGEngine
    from ..config import MemGConfig

    cfg = config or MemGConfig.from_kwargs(**kwargs)
    memg = MemGEngine(config=cfg, embedder=embedder)

    original_generate = client.generate_content

    def _patched_generate(*args: Any, **kw: Any) -> Any:
        memg._ensure_native()
        entity_id = entity or "default"

        # Normalize input
        prompt = args[0] if args else kw.get("contents", "")
        if isinstance(prompt, str):
            contents = [{"role": "user", "parts": [{"text": prompt}]}]
        elif isinstance(prompt, list):
            contents = prompt
        else:
            contents = prompt

        # Step 1: Build memory context
        user_text = _last_user_text(contents)
        context_str = ""
        if user_text:
            context_str = memg.search(entity_id, user_text, limit=10)
            if context_str and context_str.memories:
                lines = [f"- {m.content}" for m in context_str.memories]
                context_str = (
                    "[Memory Context]\n"
                    "The following are relevant memories about this user:\n"
                    + "\n".join(lines)
                    + "\n[End Memory Context]"
                )
            else:
                context_str = ""

        # Inject context into systemInstruction if available
        if context_str:
            existing_sys = kw.get("system_instruction", getattr(client, "_system_instruction", None))
            if existing_sys:
                existing_text = str(existing_sys)
                kw["system_instruction"] = existing_text + "\n\n" + context_str
            else:
                kw["system_instruction"] = context_str

        # Step 2: Call original
        if args:
            response = original_generate(*args, **kw)
        else:
            response = original_generate(**kw)

        # Step 3: Background extraction
        if extract and entity:
            try:
                resp_text = response.text if hasattr(response, "text") else ""
                if resp_text:
                    def _bg():
                        try:
                            messages = []
                            if user_text:
                                messages.append({"role": "user", "content": user_text})
                            messages.append({"role": "assistant", "content": resp_text})
                            from ..extract import run_extraction
                            llm_call = memg._make_extraction_llm_call(
                                "gemini",
                                getattr(client, "model_name", "gemini-2.5-flash"),
                                kwargs,
                            )
                            entity_uuid = memg._resolve_entity(entity_id)
                            run_extraction(
                                memg._store, entity_uuid, messages, llm_call, memg._embedder,
                            )
                        except Exception as e:
                            logger.warning("memg: background extraction failed: %s", e)

                    t = threading.Thread(target=_bg, daemon=True)
                    t.start()
            except Exception:
                pass

        return response

    client.generate_content = _patched_generate
    client._memg = memg
    client._memg_close = lambda: memg.close()
    return client


def _last_user_text(contents: Any) -> Optional[str]:
    """Extract last user text from Gemini contents."""
    if not isinstance(contents, list):
        if isinstance(contents, str):
            return contents
        return None
    for c in reversed(contents):
        if isinstance(c, dict) and c.get("role") == "user":
            parts = c.get("parts", [])
            if parts and isinstance(parts[0], dict):
                return parts[0].get("text")
    return None
